### ISSUE TITLE
[024] Verify the entire forward pass against an independent tiny oracle

### ISSUE THREAD

**Issue by 25d-228** [25d-228/gpt-2-C#25]
**Prerequisite:** #24 and #1. **Budget:** 45–60 minutes; AI may supply the explained Python oracle, not the C model. This is the forward milestone gate.

## Files and fixed fixture
Create `tools/reference_forward.py` (Python3 standard library, scalar math only), `tests/test_model_reference.c`, and `tests/fixtures/tiny/`. Use V7,maxT8,L2,H2,C4,B2,T3, input `[[0,1,2],[2,1,0]]`, already-shifted targets `[[1,2,3],[1,0,3]]`. For global registry scalar index i, assign `q=((i%17)-8)/100`; use q for all matrices/biases and1+q for LayerNorm scales. This avoids dependence on RNG implementations. Python independently computes the documented GPT-2 graph in double precision; it must never call C or consume C outputs as expected data. Compute mean loss against all six supplied targets without shifting them again, using lesson015's stable expression.

Write whitespace numeric files `params.txt`, `embedding.txt`, `block0.txt`, `block1.txt`, `final_ln.txt`, `logits.txt`, `loss.txt` with17 significant digits. Write a manifest containing config, inputs/targets, generator hash, dtype and provenance. Exclude timestamps and absolute temporary paths from deterministic fixture metadata. C loads exact counts with the testkit and compares every element, reporting tensor/index/error.

## Teach and run
```
python3 tools/reference_forward.py --out tests/fixtures/tiny
make test TEST=model_reference
make test-double TEST=model_reference
make test-reference
make test
```
Register model_reference as a fast core test using the small committed fixture, with no network/PyTorch requirement. Implement `test-reference` to regenerate in a temporary directory and compare to the committed fixture **without overwriting it**. Compare numeric contents at the double tolerance below, rather than requiring last-decimal byte identity across different libm implementations. Compare tensor names/shapes/counts, inputs/targets, formula/config and generator identity exactly. Print provenance differences clearly; do not silently replace a golden file.

Float forward atol2e-5,rtol2e-4; double atol1e-10,rtol1e-9. Include causal suffix and batch-isolation checks against both implementations. No network is needed. A known deliberate perturbation of one saved expected logit by.01 must fail the comparison helper; this verifies that the tolerance is actually enforced.

**Mutation:** transpose one C projection or omit one residual; comparison must identify the first differing intermediate, not just final loss. Restore.

**Stop:** no backward. **Exit question:** Why is comparing two C paths that call the same kernels weaker than using this independent oracle, and why should a portable float oracle not require identical text bytes?

Record largest errors and fixture provenance in `notes/step-024.md`. Do not proceed to training with a failed gate.
